### PR TITLE
Fixed building of source maps

### LINKED ISSUE #63: [INVESTIGATE] Figure out why broccoli-concat is merging sourcemaps into .js files
Currently, source maps are turned off through #39 and #25 due to a problem with source maps not being linked to some files, but the body of the source map being included in the `.js` payload. It'd be great to add full source map support back in.

### PR DESCRIPTION
**Pull request by yomexzo**
This PR fixes #63.

@minasmart , @richgilbank , @tessalt

### CODE REVIEW

File: build-lib/lib.js [hunk 1]
```diff
@@ -39,43 +39,37 @@ module.exports = function (pathConfig, env) {
...
-      sourceMapConfig: { enabled: false }
+      outputFile: `${pkg.name}.amd.js`
     });
 
     const polyFilledAmdOutput = concat(mergeTrees([amdOutput, polyfillTree]), {
       headerFiles: ['polyfills.js'],
-      inputFiles: `${pkg.name}.amd.js`,
-      outputFile: `${pkg.name}.polyfilled.amd.js`,
-      sourceMapConfig: { enabled: false }
+      inputFiles: ['**/*.js'],
```
Comment (minasmart) [inline build-lib/lib.js]: Can you verify that this globbing doesn't doesn't duplicate the polyfills file?

File: build-lib/lib.js [hunk 2]
```diff
@@ -95,12 +89,13 @@ window.ShopifyBuy = require('shopify-buy/shopify').default;
     const minifiedTree = uglifyJavaScript(funnel(tree, {
       getDestinationPath: function (path) {
         return path.replace(/\.js/, '.min.js');
-      }
+      },
+      exclude: ['**/*.map']
```
Comment (minasmart) [inline build-lib/lib.js]: Was this the line that fixed it?

File: build-lib/lib.js [hunk 3]
```diff
@@ -95,12 +89,13 @@ window.ShopifyBuy = require('shopify-buy/shopify').default;
...
     }));
 
-    const concatenatedScripts = new Licenser([mergeTrees([tree, minifiedTree])]);
+    const licensedScripts = new Licenser([mergeTrees([tree, minifiedTree], { overwrite : true })]);
 
-    tree = mergeTrees([concatenatedScripts, nodeLibOutput]);
+    tree = mergeTrees([tree, minifiedTree, licensedScripts, nodeLibOutput], { overwrite : true });
```
Comment (minasmart) [inline build-lib/lib.js]: What was the intent of this change? Getting the thrown error on overwrite indicates that your output trees collide somewhere.

**Comment by minasmart** [comment]
This still inserts source maps: 
image

**Comment by minasmart** [comment]
That was from looking at the payload that's used for examples.

File: build-lib/lib.js (hunk 1 above)
Comment (yomexzo) [inline build-lib/lib.js]: No, it doesn't duplicate. `amdOutput` will contain only one output file which is grouped along with `polyfillTree`

File: build-lib/lib.js (hunk 2 above)
Comment (yomexzo) [inline build-lib/lib.js]: Nope. The `concat` module used earlier already generates source maps. This is to tell `uglify` not to generate its own new set of source maps. If we leave it so, a duplicate file error would be thrown.

File: build-lib/lib.js (hunk 3 above)
Comment (yomexzo) [inline build-lib/lib.js]: That's correct. In this case, `Licenser` prepends the LICENSE to the top of each JS file fed into it (both minified & originals) and then generates new files with the same name.

The `overwrite` config will make the licensed JS files overwrite the non-licensed ones. Meanwhile, maps generated from `concat` above are left untouched.

**Comment by yomexzo** [comment]
@minasmart. My initial changes only affected *production* builds. 04f5c08 addresses for *non-production* builds. i.e examples and testing.

**Comment by yomexzo** [comment]
@minasmart, successfully removed the `overwrite` config by making sure the `Licenser` did not ignore other files in its input stream when building its output.

File: build-lib/licenser.js [hunk 4]
```diff
@@ -27,15 +27,16 @@ Licenser.prototype.build = function () {
...
 
     return fileAcc.concat(list);
-  }, []).filter(fileName => {
-    return fileName.match(/^.+\.js$/);
-  });
+  }, []);
 
   files.forEach(fileName => {
     const inputBuffer = fs.readFileSync(path.join(fileName));
+    var outputBuffer = inputBuffer;
```
Comment (minasmart) [inline build-lib/licenser.js]: no var. use let instead. 

```javascript
let outputBuffer;

if (match) {
  ouputBuffer = ...;
} else {
  outputBuffer = inputBuffer;
}

Comment (yomexzo) [inline build-lib/licenser.js]: thanks @minasmart .. updated.

**Comment by yomexzo** [comment]
kindly 👓 👀 @minasmart, @tessalt, @richgilbank 

I just pulled `master` into this branch to update it.

Thanks.

**Comment by tessalt** [comment]
lgtm

**Comment by lreeves** [comment]
First I've seen of a lot of this stuff but if it works and generates source maps then it seems good! Can you squash this and rebase?

**Comment by minasmart** [comment]
Hey! Could you rebase? I'd love to merge this in.

File: build-lib/lib.js [hunk 5]
```diff
@@ -108,7 +103,7 @@ window.ShopifyBuy = require('shopify-buy/shopify').default;
       ], { templateString: '{{versionString}}' })
     ]);
 
-    tree = mergeTrees([concatenatedScripts, nodeLibOutput]);
+    tree = mergeTrees([licensedAndVersionedTree, nodeLibOutput]);
```
Comment (yomexzo) [inline build-lib/lib.js]: hey @minasmart, changed this during the rebase while resolving conflict. for review before merging

File: build-lib/util/licenser.js [hunk 6]
```diff
@@ -1,4 +1,5 @@
-/* global require, module, __dirname */
+/* global require, module, fs, __dirname */
```
Comment (yomexzo) [inline build-lib/util/licenser.js]: here too @minasmart

**Comment by richgilbank** [comment]
@yomexzo Can you rebase this instead of merging master into it? Merging creates new commits each time.

**Comment by minasmart** [comment]
Yes. Please do a proper rebase, then afterward, an interactive rebase to squash commits.

**Comment by yomexzo** [comment]
good to go @minasmart , @richgilbank.

thanks @minasmart!!!

File: build-lib/util/licenser.js [hunk 7]
```diff
@@ -27,15 +28,18 @@ Licenser.prototype.build = function () {
...
-    const outputBuffer = `/*\n${LICENSE}*/\n\n${inputBuffer}`;
-
+    let outputBuffer;
+
+    if (fileName.match(/^.+\.js$/)) {
+      outputBuffer = `/*\n${LICENSE}*/\n\n${inputBuffer}`;
+    } else {
+      outputBuffer = inputBuffer;
+    }
+
```
Comment (minasmart) [inline build-lib/util/licenser.js]: There's some weird dangling whitespace on this line.

**Pull request merged**